Add output response parser

jlalmes opened the pull request:
### 🚧 Work in progress

## Added output validation to router queries & mutations

I have a goal to try and auto-generate OpenAPI docs for a TRPC router (cc. https://github.com/trpc/trpc/issues/755). The feature in this PR will enable us to generate JSON schemas for request and response payloads. This output parser is **opt-in only** and should not introduce any breaking changes.

### API
```
const appRouter = new trpc.router<Context>()
  .query({
    input: z.string(),
    output: z.object({
      input: z.string(),
    }),
    resolve: ({ input }) => {
      return { input };
    }
  });
```

- Added a new optional `output` property.
- When `output` is provided, the `resolve` function will expect a return type inferred from the `output` parser.
- The response data will be validated against the `output` parser, unless `NODE_ENV` is `production`. 
- If the response data fails validation a `TRPCError INTERNAL_SERVER_ERROR` error will be thrown.

### Todo
- Get some feedback (cc. @KATT)
- Write some comprehensive output parser tests.
- Update documentation.

KATT reviewed (commented):
Not able to look it through until tomorrow at the earliest but it is great! I would love for optional output validation

KATT commented:
I think if there is an output validation, it should always run, we can add the optionality as a feature later!

KATT commented:
Also setup github sponsors so I can send you 💷💷!

sachinraja commented:
> I have a goal to try and auto-generate OpenAPI docs for a TRPC router

Let me know if there's any way I can help! I have some experience in this area with [`zod-to-ts`](https://github.com/sachinraja/zod-to-ts) and [tRPC Playground](https://github.com/sachinraja/zod-to-ts).

KATT commented on packages/server/test/inference.test.ts:
```diff
@@ -16,6 +17,25 @@ test('infer input & output', async () => {
+      output: z.object({
+        input: z.string(),
+      }),
+      // @ts-ignore - ensuring types from "output" are expected
```
```suggestion
      // @ts-expect-error - ensuring types from "output" are expected
```

jlalmes replied:
Done 👍

KATT commented on packages/server/test/inference.test.ts:
```diff
@@ -16,6 +17,25 @@ test('infer input & output', async () => {
+    })
+    .query('withOutput', {
+      input: z.string(),
+      output: z.object({
+        input: z.string(),
+      }),
```
Maybe tests to do with this should be in it's own file? Like `output-parser.test.ts`

jlalmes replied:
I have kept these high-level type inference related tests in this file. However, I have also added `outputParser.test.ts` with a more comprehensive set of tests relating to this PR, hope this is ok. 👍

KATT commented on packages/server/src/internals/procedure.ts:
```diff
@@ -97,22 +86,26 @@ function getParseFn<TParsedInput>(
-    this.parse = getParseFn(this.inputParser);
+    this.parseInputFn = getParseFn(this.inputParser);
+    this.outputParser = opts.outputParser;
+    this.parseOutputFn = this.outputParser && getParseFn(this.outputParser);
```
This could maybe be simplified for usage it just has a passthrough fn

```suggestion
    this.parseOutputFn = this.outputParser ?? (val) => val;
```

jlalmes replied:
outdated

KATT commented on packages/server/src/internals/procedure.ts:
```diff
@@ -121,6 +114,20 @@ export class Procedure<TInputContext, TContext, TInput, TParsedInput, TOutput> {
 
+  private async parseOutput(rawOutput: unknown): Promise<TOutput> {
+    try {
+      if (process.env.NODE_ENV === 'production' || !this.parseOutputFn) {
+        return rawOutput as TOutput;
+      }
+      return await this.parseOutputFn(rawOutput);
```
Would be nice if it was:

```suggestion
  return await this.parseOutputFn(rawOutput);
```

jlalmes replied:
outdated

KATT commented on packages/server/src/internals/procedure.ts:
```diff
@@ -121,6 +114,20 @@ export class Procedure<TInputContext, TContext, TInput, TParsedInput, TOutput> {
+      }
+      return await this.parseOutputFn(rawOutput);
+    } catch (cause) {
+      throw new TRPCError({
+        code: 'INTERNAL_SERVER_ERROR',
+        cause,
+      });
```
Definitely, it's an internal server error if this is wrong. Unsure if we should do more to guide the developer.

Maybe something like this?:

```suggestion
      throw new TRPCError({
        code: 'INTERNAL_SERVER_ERROR',
        cause,
        message: 'Output validation failed',
      });
```

jlalmes replied:
done 👍

KATT reviewed (commented):
This all looks really sound to me. GOOD JOB!! 🙏 🙏 

Some suggestions:

- Make the output parser stuff into its own test file
- Don't allow bad outputs regardless of the environment
- Some smaller comments inline

KATT commented:
You might've accidentally removed the #1182 feature BTW ☺

jlalmes commented:
> #1182

Rolled it back - will make sure that any transforms in the output zod schemas are also handled!

KATT commented:
> > #1182
> 
> Rolled it back - will make sure that any transforms in the output zod schemas are also handled!

Seems like an edge case to have transformers in the output but why not!

jlalmes commented on packages/server/src/internals/procedure.ts:
```diff
@@ -100,29 +96,49 @@ function getParseFn<TParsedInput>(
+  public readonly inputParser: ProcedureParser<TParsedInput>;
+  private parseInputFn: ParseFn<TParsedInput>;
+  public readonly outputParser: ProcedureParser<TOutput> | undefined;
+  private parseOutputFn: ParseFn<TOutput>;
```
I have changed `inputParser` and `outputParser` back to `private` for now as this is out of scope for this PR. I imagine we will change them to `public` to generate openapi documentation in a future PR though 👍

jlalmes commented on packages/server/src/internals/procedure.ts:
```diff
@@ -233,19 +256,14 @@ export type CreateProcedureOptions<
 ): Procedure<unknown, TContext, TInput, TParsedInput, TOutput> {
   const inputParser =
-    'input' in opts
+    'input' in opts && opts.input
```
done 👍

jlalmes commented:
## Update

Hi @KATT & @sachinraja - I have just updated this PR & resolved all outstanding comments, please review again and leave any feedback!

### Changes to this PR
* Ensure transformers are handled correctly in zod output schemas (https://github.com/trpc/trpc/pull/1699#issuecomment-1079649264)
* Removed `NODE_ENV === "production"` output validation bypass
* Added env var `TRPC_SKIP_OUTPUT_VALIDATION === "true"` output validation bypass
* Added `"Output validation failed"` error message on validation failure
* Reverted `outputParser` & `inputParser` back to `private` class members of Procedure (outside scope of this PR)
* Added a comprehensive tests for this PR to `outputParser.test.ts` file
* Refactoring

### Outstanding tasks
* API documentation
* Add examples

.

jlalmes commented:
Just a thought - Currently `output` is omitted from `subscription()` method. 

Do you think we should allow `output` in `subscription()` and ensure that we are validating the `emit.data(output)` output payload?

KATT reviewed (commented):
Played around with this and it works great as far as I can see.

Last thing - could you add a section in the docs? `/www` folder

KATT reviewed (commented):
Will merge this tomorrow and do a release.

I removed the `TRPC_SKIP_OUTPUT_VALIDATION` because I don't want magic runtime variables - this should either be explicitly implemented per procedure or on the app router similarily to transformers/error formatters.
